query/excluder.go

```diff
@@ -0,0 +1,94 @@
+package query
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+//Excluder allows us to support different methods of excluding in the future
+type Excluder interface {
+	ShouldExclude(path string) bool
+}
+
+//RegexpExclude uses regular expressions to tell if a file/path should be excluded
+type RegexpExclude struct {
+	recursive  bool
+	exclusions []string
+	regex      string
+}
+
+//ShouldExclude will return a boolean denoting whether or not the path should be excluded based on the given slice of exclusions
+func (r *RegexpExclude) ShouldExclude(path string) bool {
+	var b bool
+	var ok error
+	if r.regex == "" {
+		r.buildRegex()
+	}
+
+	if b, ok = regexp.MatchString(r.regex, path); not(ok) {
+		fmt.Printf("regexp failed: %s \n" + ok.Error())
+		return false
+	}
+	return b
+}
+
+func (r *RegexpExclude) buildRegex() {
+	var regex string
+	var prev string
+	var curr string
+	for _, exclusion := range r.exclusions {
+		prev = curr
+		if strings.HasSuffix(exclusion, "/") {
+			curr =
+				or(mustEnd(mustBegin(escape(exclusion)+"/.*")),
+					mustEnd(mustBegin(escape(exclusion[:len(exclusion)-1]))))
+		} else {
+			curr =
+				or(mustEnd(mustBegin(escape(exclusion))),
+					mustEnd(mustBegin(escape(exclusion)+"/.*")))
+		}
+		regex = or(prev, curr)
+	}
+	r.regex = regex
```

Comment:
Since `r.regex` is of type `*regexp.Regexp` now, this should read

```go
r.regex = regexp.MustCompile(regex)
```

Note that this panics on invalid regex, which I think is the appropriate solution here, since it means the input was invalid/malformed.